src/dbnode/encoding/proto/custom_unmarshal.go

```diff
@@ -0,0 +1,381 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/dynamic"
+)
+
+var (
+	// Groups in the Protobuf wire format are deprecated, so simplify the code significantly by
+	// not supporting them.
+	errGroupsAreNotSupported = errors.New("use of groups in proto wire format is not supported")
+	zeroValue                unmarshalValue
+)
+
+type customFieldUnmarshaler interface {
+	sortedCustomFieldValues() sortedCustomFieldValues
+	nonCustomFieldValues() *dynamic.Message
+	numNonCustomValues() int
+	resetAndUnmarshal(schema *desc.MessageDescriptor, buf []byte) error
+}
+
+type customUnmarshaler struct {
+	schema *desc.MessageDescriptor
+
+	decodeBuf *buffer
+
+	customValues sortedCustomFieldValues
+
+	nonCustomValues *dynamic.Message
+	numNonCustom    int
+}
+
+func newCustomFieldUnmarshaler() customFieldUnmarshaler {
+	return &customUnmarshaler{
+		decodeBuf: newCodedBuffer(nil),
+	}
+}
+
+func (u *customUnmarshaler) sortedCustomFieldValues() sortedCustomFieldValues {
+	return u.customValues
+}
+
+func (u *customUnmarshaler) numNonCustomValues() int {
+	return u.numNonCustom
+}
+
+func (u *customUnmarshaler) nonCustomFieldValues() *dynamic.Message {
+	if u.nonCustomValues == nil {
+		u.nonCustomValues = dynamic.NewMessage(u.schema)
+	}
+	return u.nonCustomValues
+}
+
+func (u *customUnmarshaler) unmarshal() error {
+	u.customValues = u.customValues[:0]
+
+	if u.nonCustomValues != nil {
+		u.nonCustomValues.Reset()
+	}
+
+	isSorted := true
+	for !u.decodeBuf.eof() {
+		tagAndWireTypeStartOffset := u.decodeBuf.index
+		fieldNum, wireType, err := u.decodeBuf.decodeTagAndWireType()
+		if err != nil {
+			return err
+		}
+
+		fd := u.schema.FindFieldByNumber(fieldNum)
+		if fd == nil {
+			return fmt.Errorf("encountered unknown field with field number: %d", fieldNum)
+		}
+
+		isCustomField := u.isCustomField(fd)
+		if !isCustomField {
+			if u.nonCustomValues == nil {
+				u.nonCustomValues = dynamic.NewMessage(u.schema)
+			}
+
+			_, err = u.skip(wireType)
+			if err != nil {
+				return err
+			}
+
+			var (
+				startIdx       = tagAndWireTypeStartOffset
+				endIdx         = u.decodeBuf.index
+				marshaledField = u.decodeBuf.buf[startIdx:endIdx]
+			)
+			// A marshaled Protobuf message consists of a stream of <fieldNumber, wireType, value>
+			// tuples, all of which are optional, with no additional header or footer information.
+			// This means that each tuple within the stream can be thought of as its own complete
+			// marshaled message and as a result we can build up the nonCustomValues *dynamic.Message
+			// one field at a time by calling UnmarshalMerge() on sub-slices that contain a complete
+			// tuple.
+			u.nonCustomValues.UnmarshalMerge(marshaledField)
```

Comment:
Should you handle the returned `err` from this func call?

Reply:
yes thank you